### PR TITLE
WIP: admin API CDRs exports endpoint

### PR DESCRIPTION
**Pull request by gigorok**

### CODE REVIEW

File: app/jobs/worker/cdr_export_job.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    rescue => _e
+      #todo: when failed?
+      cdr_export.update!(status: 'Failed')
+    end
+
+    private
+
+    #todo: check on ransack availability
+    def raw_sql
+      Cdr::Cdr.select(cdr_export.fields.join(', ')).order('time_start desc').ransack(cdr_export.filters).result.to_sql
```
Comment (Fivell) [inline app/jobs/worker/cdr_export_job.rb]: it should be responsibility of CdrExport class, it should be possible to use  different cdr_export object that can have different raw_sql data

File: app/models/cdr_export.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  #any cdr columns
+  ALLOWED_FIELDS = Cdr::Cdr.column_names.freeze
+
+  validates_presence_of :status, :fields, :filters
+  validate do
+    # todo add requires at time_start filter
+  end
+  validate do
+    extra_fields = fields - ALLOWED_FIELDS
+    if extra_fields.any?
```
Comment (Fivell) [inline app/models/cdr_export.rb]: Imo there should be 1 abstract Export Class with minimal validation logic, to extend it and reuse with different logic

File: app/models/cdr_export.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+#
+# Table name: cdr_exports
+#
+#  id           :integer          not null, primary key
+#  status       :string           not null
+#  fields       :string           default([]), is an Array
+#  filters      :json             not null
+#  callback_url :string
+#  created_at   :datetime
+#  updated_at   :datetime
```
Comment (Fivell) [inline app/models/cdr_export.rb]: external_id uuid can be used for callback_url integration
when export ready system notify with

POST callback_url with external_id in body or something like this

File: app/models/cdr_export.rb (hunk 2 above)
Comment (gigorok) [reply app/models/cdr_export.rb]: why we need this overhead at this level?

File: app/models/cdr_export.rb (hunk 3 above)
Comment (gigorok) [reply app/models/cdr_export.rb]: could you give me some details? we were discussing with @dmitry-sinina that GET will be enough & passing ID of export record in params to notify the client

File: app/models/cdr_export.rb (hunk 2 above)
Comment (Fivell) [reply app/models/cdr_export.rb]: @gigorok 
>> to extend it and reuse with different logic

File: app/models/cdr_export.rb (hunk 3 above)
Comment (Fivell) [reply app/models/cdr_export.rb]: @gigorok  3d party system has it's own ids , uuids etc,

Comment (gigorok) [reply app/models/cdr_export.rb]: yes, thats why after making POST request from 3d party system to yeti admin API (cdr_exports endpoint) we received export ID which have to be stored on 3d party database. in this case we can match it when callback_url will be touched

**Comment by Fivell** [comment]
Such behavior is better , agree

20 марта 2018 г., 23:27 +0200, Igor Gonchar <notifications@github.com>, писал:

File: app/jobs/worker/cdr_export_job.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    def dir_path
+      Rails.configuration.yeti_web.fetch('cdr_export').fetch('dir_path').chomp('/')
+    end
+
+    def filename
+      cdr_export.id
+    end
+
+    def find_cdr_export(id)
+      @cdr_export = CdrExport.find(id)
```
Comment (senid231) [inline app/jobs/worker/cdr_export_job.rb]: at line 8 it already assigned to `@cdr_export`

File: app/jobs/worker/cdr_export_job.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
+module Worker
+  class CdrExportJob < ActiveJob::Base
+    queue_as 'cdr_export'
+
+    attr_reader :cdr_export
+
+    def perform(cdr_export_id)
+      @cdr_export = find_cdr_export(cdr_export_id)
+
+      Cdr::Cdr.connection.execute("COPY (#{cdr_export.export_sql}) TO '#{file_path}' WITH CSV HEADER;")
```
Comment (senid231) [inline app/jobs/worker/cdr_export_job.rb]: you need instance variable `@cdr_export` for file_path
why not just add method
```ruby
def file_path_for(cdr_export)
  "#{dir_path}/#{cdr_export.id}.csv"
end
```
an use it without unnecessary instance variables like this:
```ruby
cdr_export = find_cdr_export(cdr_export_id)
file_path = file_path_for(cdr_export)
Cdr::Cdr.connection.execute("COPY (#{cdr_export.export_sql}) TO '#{file_path}' WITH CSV HEADER;")
```

File: app/models/cdr_export.rb [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    'time_start_gteq',
+    'success_eq',
+    'customer_auth_external_id_eq',
+    'failed_resource_type_id_eq',
+    'src_prefix_in_contains',
+    'dst_prefix_in_contains',
+    'src_prefix_routing_contains',
+    'dst_prefix_routing_contains',
+    'customer_acc_external_id_eq',
+    'is_last_cdr_eq'
```
Comment (senid231) [inline app/models/cdr_export.rb]: maybe elements of array should be frozen as well

File: config/routes.rb [hunk 7]
```diff
@@ -80,6 +80,7 @@
           namespace :cdr do
             jsonapi_resources :cdrs, only: [:index, :show] do
             end
+            jsonapi_resources :cdr_exports, only: [:create]
```
Comment (senid231) [inline config/routes.rb]: do we have an ability to retrieve completed export file through the API?

File: db/migrate/20180320120746_create_cdr_exports.rb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+class CreateCdrExports < ActiveRecord::Migration
+  def change
+    create_table 'sys.cdr_exports' do |t|
+      t.string :status, null: false
+      t.string :fields, default: [], array: true
```
Comment (senid231) [inline db/migrate/20180320120746_create_cdr_exports.rb]: maybe `null: false` needed too?

File: config/routes.rb (hunk 7 above)
Comment (gigorok) [reply config/routes.rb]: later, in this step only create allowed

**Pull request closed without merging**